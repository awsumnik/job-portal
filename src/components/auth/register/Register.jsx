import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import styled from 'styled-components';
import { DEFAULT_ROUTES, ROLES, THEME_CONFIG } from '../../../constants';
import Button from '../../common/button/Button';
import Card from '../../common/card/Card';
import Input from '../../common/input/Input';
import { StyledCenteredWrapper } from '../../common/styled/StyledComponents';
import { handleRegisterAction } from '../../../reducers/user/actionCreator';
import messages from '../../../nls/en.json';

const BtnTextConfig = {
  [ROLES.EMPLOYER]: {
    btnText: messages.JOIN_AS_CLIENT,
  },
  [ROLES.FREELANCER]: {
    btnText: messages.JOIN_AS_FREELANCER,
  },
};

const ButtonBoxWrapper = styled.div`
  display: flex;
  column-gap: 20px;
  margin-bottom: 1rem;
`;

const ButtonBox = styled.div`
  border: 1px solid ${props => THEME_CONFIG[props.theme].borderColor};
  padding: 10px;
  border-radius: 3px;
  width: 200px;
  text-align: center;
  &.active {
    border: 1px solid ${props => THEME_CONFIG[props.theme].secondaryColor};
    background: ${props => THEME_CONFIG[props.theme].btnSecondaryBg};
    color: ${props => THEME_CONFIG[props.theme].secondaryColor};
  }
`;

const StyledLink = styled.span`
  cursor: pointer;
  color: ${props => THEME_CONFIG[props.theme].btnPrimaryColor};
`;

const Register = () => {

  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { theme, user: { loading, error, isLoggedIn: isRegistered } } = useSelector(state => state);
  const [userType, setUserType] = useState('');
  const [showForm, setShowForm] = useState(false);

  const [formErrors, setFormErrors] = useState({
    firstName: true,
    lastName: true,
    email: true,
    password: true,
  });

  const onFormError = (errObj) => {
    setFormErrors({
      ...formErrors,
      ...errObj,
    });
  };

  const hasError = () => Object.values(formErrors).some(el => el);

  const initialFormData = Object.freeze({
    firstName: '',
    lastName: '',
    email: '',
    password: '',
  });

  const [formData, updateFormData] = useState(initialFormData);

  const handleChange = (e) => {
    updateFormData({
      ...formData,
      [e.target.name]: e.target.value.trim()
    });
  };

  useEffect(() => {
    if (isRegistered) {
      const redirectPath = DEFAULT_ROUTES[userType];
      navigate(redirectPath);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isRegistered]);

  const handleRegisterClick = (event) => {
    event.preventDefault();
    const payload = {
      id: uuidv4(),
      ...formData,
      userType,
      password: btoa(formData.password),
    };
    dispatch(handleRegisterAction(payload));
  };

  const switchUserType = (userType) => {
    setShowForm(false);
    setUserType(userType);
  };

  const goToSignIn = () => {
    navigate('/sign-in');
  };

  const renderRegisterForm = () => {
    return (
      <>
        <form className='form' autoComplete='off' onSubmit={handleRegisterClick}>
          <div>
            <Input
              type="text"
              label="First Name"
              name="firstName"
              placeholder={'Enter First Name'}
              onChange={handleChange}
              isRequired
              autoFocus
              minLength={3}
              onError={onFormError}
            />
          </div>

          <div>
            <Input
              type="text"
              label="Last Name"
              name="lastName"
              placeholder={'Enter Last Name'}
              onChange={handleChange}
              isRequired
              onError={onFormError}
            />
          </div>

          {
            userType === ROLES.EMPLOYER && (
              <div>
                <Input
                  type="text"
                  label="Company Name"
                  name="companyName"
                  placeholder={'Enter Company name'}
                  onChange={handleChange}
                  isRequired
                  minLength={2}
                />
              </div>
            )
          }

          <div>
            <Input
              type="email"
              label="Email"
              name="email"
              placeholder={'Enter Email'}
              onChange={handleChange}
              isRequired
              isEmail
              minLength={8}
              onError={onFormError}
            />
          </div>

          <div>
            <Input
              type="password"
              label="Password"
              name="password"
              placeholder={'Enter Password'}
              onChange={handleChange}
              isRequired
              isPassword
              minLength={8}
              onError={onFormError}
            />
          </div>

          <div>
            <Button
              type="submit"
              label={messages.CREATE_MY_ACCOUNT}
              name="registerBtn"
              priority="primary"
              onClick={handleRegisterClick}
              isDisabled={hasError()}
            />
          </div>

          <div className='text-secondary text-center'>
            {
              userType === ROLES.FREELANCER ? (
                <>
                  {messages.HIRE_TALENT}
                  <StyledLink theme={theme} onClick={() => {
                    switchUserType(ROLES.EMPLOYER)
                  }} >{messages.JOIN_AS_CLIENT}</StyledLink>
                </>
              ) : (
                <>
                  {messages.LOOKING_FOR_WORK}
                  <StyledLink theme={theme} onClick={() => {
                    switchUserType(ROLES.FREELANCER)
                  }} >{messages.APPLY_AS_TALENT}</StyledLink>
                </>
              )

            }
          </div>
        </form>
      </>
    )
  }

  return (
    <StyledCenteredWrapper>
      <Card style={{ width: 400 }}>
        <div className="card-title text-center">{messages.JOIN_US}</div>
        {
          error && !loading && (
            <div className='error-text text-center'>{error}</div>
          )
        }
        <div className='card-content'>
          {
            showForm ? renderRegisterForm() : (
              <>
                <ButtonBoxWrapper>
                  <ButtonBox theme={theme} className={userType===ROLES.EMPLOYER ? 'active': ''} onClick={() => { setUserType(ROLES.EMPLOYER) }}>
                    {messages.IAM_CLIENT}
                  </ButtonBox>
                  <ButtonBox theme={theme} className={userType===ROLES.FREELANCER ? 'active': ''} onClick={() => { setUserType(ROLES.FREELANCER) }}>
                    {messages.IAM_FREELANCER}
                  </ButtonBox>
                </ButtonBoxWrapper>
                <div>
                  <Button
                    type="button"
                    label={BtnTextConfig[userType]?.btnText || messages.CREATE_ACCOUNT}
                    name="joinBtn"
                    priority="primary"
                    isDisabled={!userType}
                    onClick={() => { setShowForm(true) }}
                  />
                </div>
              </>
            )
          }
          <div>
            <hr />
            <p className='text-center'>{messages.ALREADY_HAVE_ACCOUNT}</p>
            <Button
              type="button"
              label="Log In"
              name="signInBtn"
              priority="secondary"
              onClick={goToSignIn}
            />
          </div>
        </div>
      </Card>
    </StyledCenteredWrapper>
  )
}

export default Register;