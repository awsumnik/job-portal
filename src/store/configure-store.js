import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';
import { loadInititalAppState, savedState, saveState } from './';
loadInititalAppState()

let store;
if (process.env.NODE_ENV === 'development') {
  store = createStore(
    rootReducer,
    savedState,
    composeWithDevTools(applyMiddleware(thunk)),
  );
} else {
  store = createStore(rootReducer, savedState, applyMiddleware(thunk));
}

store.subscribe(() => {
  saveState(store.getState());
});

export default store;