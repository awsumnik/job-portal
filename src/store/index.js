import { themeIntitialState } from '../reducers/theme/themeReducer';
import { userInitialState } from '../reducers/user/userReducer';

const initialState = {
  theme: themeIntitialState,
  user: userInitialState,
};

export const loadState = () => {
  const state = localStorage.getItem('job_portal_state');
  if (state === null) {
    return undefined;
  }
  return JSON.parse(state);
};

export const saveState = (state) => {
  const { user, theme } = state;
  const currentState = JSON.stringify({ theme, user });
  localStorage.setItem('job_portal_state', currentState);
};

export const savedState = loadState();

export const loadInititalAppState = () => {
  saveState(initialState)
}