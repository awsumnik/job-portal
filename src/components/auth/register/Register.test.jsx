import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Register from './Register';
import store from '../../../store/configure-store';

describe('Register Page', () => {

  test('should render register page', async () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
    const linkElement = screen.getByText(/Join as a client or freelancer/i);
    expect(linkElement).toBeInTheDocument();
    const registerButton = screen.getByRole('button', {
      name: /create account/i
    });
    expect(registerButton).toBeDisabled();
    const joinAsFreelancerBtn = screen.getByText(/I'm a client, hiring for a project/i);
    fireEvent.click(joinAsFreelancerBtn);

    await waitFor(() => {
      expect(joinAsFreelancerBtn).toBeEnabled();
    })
  });

  test('should display error for invalid input', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );
    const joinAsClientBtn = screen.getByText(/I'm a client, hiring for a project/i);
    fireEvent.click(joinAsClientBtn);
    await waitFor(() => {
      expect(joinAsClientBtn).toBeEnabled();
    });

    const registerButton = screen.getByRole('button', {
      name: /join as a client/i
    });
    fireEvent.click(registerButton);

    const emailInput = screen.getByPlaceholderText('Enter Email');
    fireEvent.change(emailInput, { target: { value: 'test@123' } });
    fireEvent.blur(emailInput);
    expect(screen.getByText(/Enter a valid Email./i)).toBeInTheDocument();
  });

  test('register button to be enabled after valid inputs are entered', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Register />
        </Router>
      </Provider>
    );
    const joinAsFreelancerBtn = screen.getByText(/I'm a freelancer, looking for work/i);
    fireEvent.click(joinAsFreelancerBtn);
    await waitFor(() => {
      expect(joinAsFreelancerBtn).toBeEnabled();
    });

    const registerButton = screen.getByRole('button', {
      name: /apply as a freelancer/i
    });
    fireEvent.click(registerButton);

    const emailInput = screen.getByPlaceholderText('Enter Email');
    const passwordInput = screen.getByPlaceholderText('Enter Password');
    const fNameInput = screen.getByPlaceholderText('Enter First Name');
    const lNameInput = screen.getByPlaceholderText('Enter Last Name');
    fireEvent.change(fNameInput, { target: { value: 'Anil' } });
    expect(fNameInput.value).toBe('Anil');
    fireEvent.blur(fNameInput);
    fireEvent.change(lNameInput, { target: { value: 'Saini' } });
    expect(lNameInput.value).toBe('Saini');
    fireEvent.blur(lNameInput);
    fireEvent.change(emailInput, { target: { value: 'anil@test.com' } });
    expect(emailInput.value).toBe('anil@test.com');
    fireEvent.blur(emailInput);
    fireEvent.change(passwordInput, { target: { value: 'Test-123' } });
    expect(emailInput.value).toBe('anil@test.com');
    fireEvent.blur(passwordInput);

    const regiterNowButton = screen.getByRole('button', {
      name: /create my account/i
    });
    await waitFor(() => {
      expect(regiterNowButton).toBeEnabled();
    });
    fireEvent.click(regiterNowButton);
  });
});