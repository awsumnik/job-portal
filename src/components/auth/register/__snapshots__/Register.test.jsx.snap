// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render register page 1`] = `
<div>
  <div
    class="sc-dkzDqf iGZwPq"
  >
    <div
      class="sc-gsnTZi eyomTB card"
      style="width: 400px;"
    >
      <div
        class="card-title text-center"
      >
        Join as a client or freelancer
      </div>
      <div
        class="card-content"
      >
        <div
          class="sc-ftvSup fvORvb"
        >
          <div
            class="sc-papXJ ceQOAN"
          >
            I'm a client, hiring for a project
          </div>
          <div
            class="sc-papXJ ceQOAN"
          >
            I'm a freelancer, looking for work
          </div>
        </div>
        <div>
          <div
            style="display: flex; justify-content: center;"
          >
            <button
              class="sc-bczRLJ cVLaUc form-button"
              disabled=""
              id="joinBtn"
              name="joinBtn"
              type="button"
            >
              Create Account
            </button>
          </div>
        </div>
        <div>
          <hr />
          <p
            class="text-center"
          >
            Already have an account?
          </p>
          <div
            style="display: flex; justify-content: center;"
          >
            <button
              class="sc-bczRLJ cBkcFt form-button"
              id="signInBtn"
              name="signInBtn"
              type="button"
            >
              Log In
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
